protocols/upnp/src/lib.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of UPnP port mapping for libp2p.
+//!
+//! This crate provides a `tokio::Behaviour` and `async_std::Behaviour`, depending on the enabled features, which
+//! implement the [`libp2p_swarm::NetworkBehaviour`] trait. This struct will automatically try to map the ports externally to internal
+//! addresses on the gateway.
+//!
+
+#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
+
+mod behaviour;
+mod provider;
+
+use std::{
+    net::{Ipv4Addr, SocketAddr, SocketAddrV4},
+    time::Duration,
+};
+
+pub use behaviour::Event;
+
+#[cfg(feature = "async-std")]
+pub use provider::async_std;
+#[cfg(feature = "tokio")]
+pub use provider::tokio;
+
+/// The configuration for UPnP capabilities for libp2p.
+#[derive(Clone, Copy, Debug)]
+pub struct Config {
+    /// Bind address for UDP socket (defaults to all `0.0.0.0`).
+    bind_addr: SocketAddr,
+    /// Broadcast address for discovery packets (defaults to `239.255.255.250:1900`).
+    broadcast_addr: SocketAddr,
+    /// Timeout for a search iteration (defaults to 10s).
+    timeout: Option<Duration>,
+    /// Should the port mappings be temporary (1 hour) or permanent.
+    temporary: bool,
+}
+
+impl Config {
+    /// Creates a new configuration for a UPnP transport:
+    pub fn new() -> Self {
+        Self {
+            bind_addr: SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), 0)),
+            broadcast_addr: "239.255.255.250:1900".parse().unwrap(),
+            timeout: Some(Duration::from_secs(10)),
+            temporary: true,
+        }
+    }
+
+    /// Configures the port mappings to be temporary (1 hour) or permanent.
+    pub fn temporary(self, temporary: bool) -> Self {
+        Self { temporary, ..self }
+    }
```

Comment:
I am still not sure we should actually expose this. Wouldn't a temporary mapping always be better?

Reply:
If we don't have this one, do we even need a `Config`? :)

We could just implement `Default`. If we ever want to configure something later, we can add a `new` function to the behaviour in a backwards-compatible way.

Reply:
My guess would be that most developers using this API are not going to be the users of this software. In other words, they don't know whether the target router will support this mapping. Yet, with this configuration option, they still have to make a decision on whether to use it or not. How should they make this decision?

I think we have two options:

- Always only try temporary and if it fails, don't do anything
- Try temporary and fall back to a permanent one if that fails

The 2nd one is a bit more dangerous although I am not sure how much it matters in practice. What does the go implementation do here?

The safe bet is IMO to simply fail if temporary mappings are not supported. I'd suggest we start with that. We can always ship a feature later to use permanent mappings for some cases.

In any case, I don't think this configuration option adds much value.

Reply:
Sorry Thomas, missed this comment. 

>My guess would be that most developers using this API are not going to be the users of this software. In other words, they don't know whether the target router will support this mapping. Yet, with this configuration option, they still have to make a decision on whether to use it or not. How should they make this decision?

yeah that makes sense. Updated to remove the `Config` ptal :)

> The 2nd one is a bit more dangerous although I am not sure how much it matters in practice. What does the go implementation do here?

yeah, I wouldn't also go for it as it don't have `AsyncDrop` yet and it's not possible to remove the mappings on `Drop` meaning they become mapped forever.

from what I read it also seems the [go implementation doesn't support permanent mappings](https://github.com/libp2p/go-libp2p/blob/c02be3228d9536aac7be39e05f729e9534778206/p2p/net/nat/nat.go#L116-L120)?